pymbolic/primitives.py

```diff
@@ -20,10 +22,17 @@
 THE SOFTWARE.
 """
 
+from typing import Sequence, Union, Mapping, Type, Optional, Callable, Any, Tuple
 from sys import intern
-from abc import ABC, abstractmethod
+from abc import abstractmethod
 import pymbolic.traits as traits
 
+import attrs
+
+
+_ConstantT = Union[int, float, complex]
+ExpressionT = Union[_ConstantT, "Expression", Tuple["ExpressionT", ...]]
```

Comment:
These are a bit of a lie, in that, at least in theory, all sorts of constants are allowed. (numpy arrays, what have you) I don't quite know what to do about that in terms of typing though. We could:

- Throw an `Any` in there :thinking: 
- Just live with the *cough* underapproximation and judiciously type-ignore in user packages.

Reply:
To add to those:
* Throw an `np.generic` in there?
* Some form of protocol? Like those [SupportsInt](https://github.com/python/cpython/blob/c00eb1eae6af3ee5b7e314add4606da4521bb8c5/Lib/typing.py#L2750) things, but there probably isn't one that makes sense..

Reply:
> Throw an `np.generic` in there?

Can't! We don't have a numpy dependency here, and I'm not eager to add one. And for type annotations, we cannot conditionally *not* import AFAIK.

> Some form of protocol? Like those `SupportsInt` things, but there probably isn't one that makes sense..

Well, `Number` might make sense, but mypy has just [given up on doing any such thing](https://github.com/python/mypy/issues/3186). A corresponding [discussion](https://discuss.python.org/t/numeric-generics-where-do-we-go-from-pep-3141-and-present-day-mypy/17155/14) also seems to have petered out. I think judiciously ignoring abuses of this is probably the way to go.